File: Zombie_Killer/core/src/com/nopalsoft/zombiekiller/game/TiledMapManagerBox2d.java
package com.nopalsoft.zombiekiller.game;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.maps.Map;
import com.badlogic.gdx.maps.MapLayer;
import com.badlogic.gdx.maps.MapObject;
import com.badlogic.gdx.maps.MapObjects;
import com.badlogic.gdx.maps.MapProperties;
import com.badlogic.gdx.maps.objects.CircleMapObject;
import com.badlogic.gdx.maps.objects.PolygonMapObject;
import com.badlogic.gdx.maps.objects.PolylineMapObject;
import com.badlogic.gdx.maps.objects.RectangleMapObject;
import com.badlogic.gdx.maps.objects.TextureMapObject;
import com.badlogic.gdx.math.Circle;
import com.badlogic.gdx.math.Rectangle;
import com.badlogic.gdx.math.Vector2;
import com.badlogic.gdx.physics.box2d.Body;
import com.badlogic.gdx.physics.box2d.BodyDef;
import com.badlogic.gdx.physics.box2d.BodyDef.BodyType;
import com.badlogic.gdx.physics.box2d.ChainShape;
import com.badlogic.gdx.physics.box2d.CircleShape;
import com.badlogic.gdx.physics.box2d.FixtureDef;
import com.badlogic.gdx.physics.box2d.PolygonShape;
import com.badlogic.gdx.physics.box2d.Shape;
import com.badlogic.gdx.physics.box2d.World;
import com.badlogic.gdx.utils.GdxRuntimeException;
import com.badlogic.gdx.utils.Logger;
import com.nopalsoft.zombiekiller.game_objects.Crate;
import com.nopalsoft.zombiekiller.game_objects.ItemGem;
import com.nopalsoft.zombiekiller.game_objects.ItemHeart;
import com.nopalsoft.zombiekiller.game_objects.ItemMeat;
import com.nopalsoft.zombiekiller.game_objects.ItemShield;
import com.nopalsoft.zombiekiller.game_objects.ItemSkull;
import com.nopalsoft.zombiekiller.game_objects.ItemStar;
import com.nopalsoft.zombiekiller.game_objects.Items;
import com.nopalsoft.zombiekiller.game_objects.Platform;
import com.nopalsoft.zombiekiller.game_objects.Saw;
import com.nopalsoft.zombiekiller.game_objects.Zombie;

import java.util.Iterator;

public class TiledMapManagerBox2d {

    private final WorldGame worldGame;
    private final World world;
    private final float m_units;
    private final Logger logger;
    private final FixtureDef defaultFixtureDefinition;

    public TiledMapManagerBox2d(WorldGame worldGame, float unitScale) {
        this.worldGame = worldGame;
        world = worldGame.world;
        m_units = unitScale;
        logger = new Logger("MapBodyManager", 1);

        defaultFixtureDefinition = new FixtureDef();
        defaultFixtureDefinition.density = 1.0f;
        defaultFixtureDefinition.friction = .5f;
        defaultFixtureDefinition.restitution = 0.0f;
    }

    public void createObjectsFromTiled(Map map) {
        createPhysics(map);
        createEnemies(map);
    }

    private void createPhysics(Map map) {
        MapLayer layer = map.getLayers().get("fisicos");

        if (layer == null) {
            logger.error("layer " + "fisicos" + " no existe");
            return;
        }

        MapObjects objects = layer.getObjects();
        Iterator<MapObject> objectIterator = objects.iterator();

        int skulls = 0;
        while (objectIterator.hasNext()) {
            MapObject object = objectIterator.next();

            if (object instanceof TextureMapObject) {
                continue;
            }

            MapProperties properties = object.getProperties();
            String type = (String) properties.get("type");
            if (type == null)
                continue;
            else if (type.equals("pisable")) {
                if (object instanceof RectangleMapObject) {
                    createPlatform(object);
                    continue;
                }
            } else if (type.equals("ladder")) {
                if (object instanceof RectangleMapObject) {
                    createLadder(object, type);
                    continue;
                }
            } else if (type.equals("crate")) {
                if (object instanceof RectangleMapObject) {
                    createCrate(object);
                    continue;
                }
            } else if (type.equals("saw")) {
                if (object instanceof RectangleMapObject) {
                    createSaw(object);
                    continue;
                }
            } else if (type.equals("gem") || type.equals("heart") || type.equals("star") || type.equals("skull") || type.equals("meat")
                    || type.equals("shield")) {
                if (object instanceof RectangleMapObject) {
                    if (type.equals("skull"))
                        skulls++;
                    createItem(object, type);
                    continue;
                }
            }

            /*
             * Normally if not none is the floor.
             */
            Shape shape;
            if (object instanceof RectangleMapObject) {
                shape = getRectangle((RectangleMapObject) object);
            } else if (object instanceof PolygonMapObject) {
                shape = getPolygon((PolygonMapObject) object);
            } else if (object instanceof PolylineMapObject) {
                shape = getPolyline((PolylineMapObject) object);
            } else if (object instanceof CircleMapObject) {
                shape = getCircle((CircleMapObject) object);
            } else {
                logger.error("non suported shape " + object);
                continue;
            }

            defaultFixtureDefinition.shape = shape;

            BodyDef bodyDefinition = new BodyDef();
            bodyDefinition.type = BodyType.StaticBody;

            Body body = world.createBody(bodyDefinition);
            body.createFixture(defaultFixtureDefinition);
            body.setUserData(type);

            defaultFixtureDefinition.shape = null;
            defaultFixtureDefinition.isSensor = false;
            defaultFixtureDefinition.filter.groupIndex = 0;
            shape.dispose();
        }
        if (skulls != 3)
            throw new GdxRuntimeException("#### DEBE HABER 3 SKULLS ####");
    }

    private void createEnemies(Map map) {
        MapLayer layer = map.getLayers().get("malos");

        if (layer == null) {
            logger.error("layer " + "malos" + " no existe");
            return;
        }

        MapObjects objects = layer.getObjects();

        for (MapObject object : objects) {
            if (object instanceof TextureMapObject) {
                continue;
            }

            MapProperties properties = object.getProperties();
            String type = (String) properties.get("type");
            if (type.equals("zombieCuasy") || type.equals("zombieFrank") || type.equals("zombieMummy") || type.equals("zombieKid")
                    || type.equals("zombiePan")) {
                if (object instanceof RectangleMapObject) {
                    createZombie(object, type);
                    worldGame.TOTAL_ZOMBIES_LEVEL++;
                }
            } else {
                throw new GdxRuntimeException("Error en layer " + "malos" + ", objeto:" + type);
            }
        }
    }

    private void createPlatform(MapObject object) {
        Rectangle rectangle = ((RectangleMapObject) object).getRectangle();
        PolygonShape polygon = new PolygonShape();
        Vector2 size = new Vector2((rectangle.x + rectangle.width * 0.5f) * m_units, (rectangle.y + rectangle.height * 0.5f) * m_units);
        polygon.setAsBox(rectangle.getWidth() * 0.5f * m_units, rectangle.height * 0.5f * m_units, size, 0.0f);
        defaultFixtureDefinition.shape = polygon;

        BodyDef bodyDef = new BodyDef();
        bodyDef.type = BodyType.StaticBody;
        Body body = world.createBody(bodyDef);
        body.createFixture(defaultFixtureDefinition);

        float x = (rectangle.x + rectangle.width * 0.5f) * m_units;
        float y = (rectangle.y + rectangle.height * 0.5f) * m_units;
        float height = (rectangle.height * m_units * 0.5f);
        float width = (rectangle.width * m_units * 0.5f);
        body.setUserData(new Platform(x, y, width, height));
    }

    private void createLadder(MapObject object, String tipo) {
        Rectangle rectangle = ((RectangleMapObject) object).getRectangle();
        PolygonShape polygon = new PolygonShape();
        Vector2 size = new Vector2((rectangle.x + rectangle.width * 0.5f) * m_units, (rectangle.y + rectangle.height * 0.5f) * m_units);
        polygon.setAsBox(rectangle.getWidth() * 0.5f * m_units, rectangle.height * 0.5f * m_units, size, 0.0f);
        defaultFixtureDefinition.shape = polygon;

        BodyDef bodyDef = new BodyDef();
        bodyDef.type = BodyType.StaticBody;
        Body body = world.createBody(bodyDef);

        defaultFixtureDefinition.isSensor = true;
        body.createFixture(defaultFixtureDefinition);
        body.setUserData(tipo);

        defaultFixtureDefinition.isSensor = false;
    }

    private void createCrate(MapObject object) {
        Rectangle rectangle = ((RectangleMapObject) object).getRectangle();

        PolygonShape polygon = new PolygonShape();
        float width = (rectangle.width * m_units);
        float x = (rectangle.x + rectangle.width * 0.5f) * m_units;
        float y = (rectangle.y + rectangle.height * 0.5f) * m_units;

        polygon.setAsBox(width / 2f, width / 2f);

        FixtureDef fixDef = new FixtureDef();
        fixDef.density = 12f;
        fixDef.friction = .7f;
        fixDef.restitution = 0.0f;
        fixDef.shape = polygon;

        Crate obj = new Crate(x, y, width);

        BodyDef bodyDefinition = new BodyDef();
        bodyDefinition.position.x = obj.position.x;
        bodyDefinition.position.y = obj.position.y;
        bodyDefinition.type = BodyType.DynamicBody;
        Body body = world.createBody(bodyDefinition);

        body.createFixture(fixDef);

        worldGame.crates.add(obj);
        body.setUserData(obj);
    }

    private void createSaw(MapObject object) {
        Rectangle rectangle = ((RectangleMapObject) object).getRectangle();

        float width = (rectangle.width * m_units);
        float x = (rectangle.x + rectangle.width * 0.5f) * m_units;
        float y = (rectangle.y + rectangle.height * 0.5f) * m_units;

        CircleShape shape = new CircleShape();
        shape.setRadius(width / 2f);

        defaultFixtureDefinition.shape = shape;

        Saw obj = new Saw(x, y, width);

        BodyDef bodyDef = new BodyDef();
        bodyDef.position.x = obj.position.x;
        bodyDef.position.y = obj.position.y;
        bodyDef.type = BodyType.KinematicBody;
        Body body = world.createBody(bodyDef);

        body.createFixture(defaultFixtureDefinition);

        worldGame.saws.add(obj);
        body.setUserData(obj);
        body.setAngularVelocity((float) Math.toRadians(360));

        shape.dispose();
    }

    private void createItem(MapObject object, String type) {
        Items obj = null;

        Rectangle rectangle = ((RectangleMapObject) object).getRectangle();
        float x = (rectangle.x + rectangle.width * 0.5f) * m_units;
        float y = (rectangle.y + rectangle.height * 0.5f) * m_units;

        switch (type) {
            case "gem":
                obj = new ItemGem(x, y);
                break;
            case "heart":
                obj = new ItemHeart(x, y);
                Gdx.app.log("Heart", "");
                break;
            case "skull":
                obj = new ItemSkull(x, y);
                break;
            case "meat":
                obj = new ItemMeat(x, y);
                break;
            case "shield":
                obj = new ItemShield(x, y);
                break;
            case "star":
                obj = new ItemStar(x, y);
                break;
        }

        BodyDef bodyDefinition = new BodyDef();
        bodyDefinition.position.y = obj.position.y;
        bodyDefinition.position.x = obj.position.x;
        bodyDefinition.type = BodyType.StaticBody;

        CircleShape shape = new CircleShape();
        shape.setRadius(.15f);

        FixtureDef fixtureDefinition = new FixtureDef();
        fixtureDefinition.shape = shape;
        fixtureDefinition.isSensor = true;

        Body body = world.createBody(bodyDefinition);
        body.createFixture(fixtureDefinition);

        body.setUserData(obj);
        worldGame.items.add(obj);
        shape.dispose();
    }

    private void createZombie(MapObject object, String type) {
        Zombie obj = null;

        Rectangle rectangle = ((RectangleMapObject) object).getRectangle();
        float x = (rectangle.x + rectangle.width * 0.5f) * m_units;
        float y = (rectangle.y + rectangle.height * 0.5f) * m_units;

        switch (type) {
            case "zombieCuasy":
                obj = new Zombie(x, y, Zombie.TIPO_CUASY);
                break;
            case "zombieFrank":
                obj = new Zombie(x, y, Zombie.TIPO_FRANK);
                break;
            case "zombieKid":
                obj = new Zombie(x, y, Zombie.TIPO_KID);
                break;
            case "zombieMummy":
                obj = new Zombie(x, y, Zombie.TIPO_MUMMY);
                break;
            case "zombiePan":
                obj = new Zombie(x, y, Zombie.TIPO_PAN);
                break;
        }

        BodyDef bodyDefinition = new BodyDef();
        bodyDefinition.position.x = obj.position.x;
        bodyDefinition.position.y = obj.position.y;
        bodyDefinition.type = BodyType.DynamicBody;

        Body body = world.createBody(bodyDefinition);

        PolygonShape shape = new PolygonShape();
        shape.setAsBox(.17f, .32f);

        FixtureDef fixtureDefinition = new FixtureDef();
        fixtureDefinition.shape = shape;
        fixtureDefinition.density = 8;
        fixtureDefinition.friction = 0;
        fixtureDefinition.filter.groupIndex = -1;
        body.createFixture(fixtureDefinition);

        body.setFixedRotation(true);
        body.setUserData(obj);
        worldGame.zombies.add(obj);

        shape.dispose();
    }

    private Shape getRectangle(RectangleMapObject rectangleObject) {
        Rectangle rectangle = rectangleObject.getRectangle();
        PolygonShape polygon = new PolygonShape();
        Vector2 size = new Vector2((rectangle.x + rectangle.width * 0.5f) * m_units, (rectangle.y + rectangle.height * 0.5f) * m_units);
        polygon.setAsBox(rectangle.getWidth() * 0.5f * m_units, rectangle.height * 0.5f * m_units, size, 0.0f);
        return polygon;
    }

    private Shape getCircle(CircleMapObject circleObject) {
        Circle circle = circleObject.getCircle();
        CircleShape circleShape = new CircleShape();
        circleShape.setRadius(circle.radius * m_units);
        circleShape.setPosition(new Vector2(circle.x * m_units, circle.y * m_units));
        return circleShape;
    }

    private Shape getPolygon(PolygonMapObject polygonObject) {
        PolygonShape polygon = new PolygonShape();
        float[] vertices = polygonObject.getPolygon().getVertices();
        float[] worldVertices = new float[vertices.length];
        float yLost = polygonObject.getPolygon().getY() * m_units;
        float xLost = polygonObject.getPolygon().getX() * m_units;

        for (int i = 0; i < vertices.length; ++i) {
            if (i % 2 == 0)
                worldVertices[i] = vertices[i] * m_units + xLost;
            else
                worldVertices[i] = vertices[i] * m_units + yLost;
        }
        polygon.set(worldVertices);

        return polygon;
    }

    private Shape getPolyline(PolylineMapObject polylineObject) {
        float[] vertices = polylineObject.getPolyline().getVertices();

        Vector2[] worldVertices = new Vector2[vertices.length / 2];
        float yLost = polylineObject.getPolyline().getY() * m_units;
        float xLost = polylineObject.getPolyline().getX() * m_units;
        for (int i = 0; i < vertices.length / 2; ++i) {
            worldVertices[i] = new Vector2();
            worldVertices[i].x = vertices[i * 2] * m_units + xLost;
            worldVertices[i].y = vertices[i * 2 + 1] * m_units + yLost;
        }
        ChainShape chain = new ChainShape();
        chain.createChain(worldVertices);
        return chain;
    }
}

File: Zombie_Killer/core/src/com/nopalsoft/zombiekiller/game/TiledMapManagerBox2d.kt
package com.nopalsoft.zombiekiller.game

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.maps.Map
import com.badlogic.gdx.maps.MapObject
import com.badlogic.gdx.maps.objects.CircleMapObject
import com.badlogic.gdx.maps.objects.PolygonMapObject
import com.badlogic.gdx.maps.objects.PolylineMapObject
import com.badlogic.gdx.maps.objects.RectangleMapObject
import com.badlogic.gdx.maps.objects.TextureMapObject
import com.badlogic.gdx.math.Vector2
import com.badlogic.gdx.physics.box2d.BodyDef
import com.badlogic.gdx.physics.box2d.BodyDef.BodyType
import com.badlogic.gdx.physics.box2d.ChainShape
import com.badlogic.gdx.physics.box2d.CircleShape
import com.badlogic.gdx.physics.box2d.FixtureDef
import com.badlogic.gdx.physics.box2d.PolygonShape
import com.badlogic.gdx.physics.box2d.Shape
import com.badlogic.gdx.physics.box2d.World
import com.badlogic.gdx.utils.GdxRuntimeException
import com.badlogic.gdx.utils.Logger
import com.nopalsoft.zombiekiller.game_objects.Crate
import com.nopalsoft.zombiekiller.game_objects.ItemGem
import com.nopalsoft.zombiekiller.game_objects.ItemHeart
import com.nopalsoft.zombiekiller.game_objects.ItemMeat
import com.nopalsoft.zombiekiller.game_objects.ItemShield
import com.nopalsoft.zombiekiller.game_objects.ItemSkull
import com.nopalsoft.zombiekiller.game_objects.ItemStar
import com.nopalsoft.zombiekiller.game_objects.Items
import com.nopalsoft.zombiekiller.game_objects.Platform
import com.nopalsoft.zombiekiller.game_objects.Saw
import com.nopalsoft.zombiekiller.game_objects.Zombie

class TiledMapManagerBox2d(worldGame: WorldGame, unitScale: Float) {
    private val worldGame: WorldGame?
    private val world: World
    private val m_units: Float
    private val logger: Logger
    private val defaultFixtureDefinition: FixtureDef

    init {
        this.worldGame = worldGame
        world = worldGame.world
        m_units = unitScale
        logger = Logger("MapBodyManager", 1)

        defaultFixtureDefinition = FixtureDef()
        defaultFixtureDefinition.density = 1.0f
        defaultFixtureDefinition.friction = .5f
        defaultFixtureDefinition.restitution = 0.0f
    }

    fun createObjectsFromTiled(map: Map) {
        createPhysics(map)
        createEnemies(map)
    }

    private fun createPhysics(map: Map) {
        val layer = map.layers.get("fisicos")

        if (layer == null) {
            logger.error("layer " + "fisicos" + " no existe")
            return
        }

        val objects = layer.objects
        val objectIterator = objects.iterator()

        var skulls = 0
        while (objectIterator.hasNext()) {
            val `object` = objectIterator.next()

            if (`object` is TextureMapObject) {
                continue
            }

            val properties = `object`.properties
            val type = properties.get("type") as String?
            if (type == null) continue
            else if (type == "pisable") {
                if (`object` is RectangleMapObject) {
                    createPlatform(`object`)
                    continue
                }
            } else if (type == "ladder") {
                if (`object` is RectangleMapObject) {
                    createLadder(`object`, type)
                    continue
                }
            } else if (type == "crate") {
                if (`object` is RectangleMapObject) {
                    createCrate(`object`)
                    continue
                }
            } else if (type == "saw") {
                if (`object` is RectangleMapObject) {
                    createSaw(`object`)
                    continue
                }
            } else if (type == "gem" || type == "heart" || type == "star" || type == "skull" || type == "meat"
                || type == "shield"
            ) {
                if (`object` is RectangleMapObject) {
                    if (type == "skull") skulls++
                    createItem(`object`, type)
                    continue
                }
            }

            /*
             * Normally if not none is the floor.
             */
            val shape: Shape?
            if (`object` is RectangleMapObject) {
                shape = getRectangle(`object`)
            } else if (`object` is PolygonMapObject) {
                shape = getPolygon(`object`)
            } else if (`object` is PolylineMapObject) {
                shape = getPolyline(`object`)
            } else if (`object` is CircleMapObject) {
                shape = getCircle(`object`)
            } else {
                logger.error("non suported shape " + `object`)
                continue
            }

            defaultFixtureDefinition.shape = shape

            val bodyDefinition = BodyDef()
            bodyDefinition.type = BodyType.StaticBody

            val body = world.createBody(bodyDefinition)
            body.createFixture(defaultFixtureDefinition)
            body.userData = type

            defaultFixtureDefinition.shape = null
            defaultFixtureDefinition.isSensor = false
            defaultFixtureDefinition.filter.groupIndex = 0
            shape.dispose()
        }
        if (skulls != 3) throw GdxRuntimeException("#### DEBE HABER 3 SKULLS ####")
    }

    private fun createEnemies(map: Map) {
        val layer = map.layers.get("malos")

        if (layer == null) {
            logger.error("layer " + "malos" + " no existe")
            return
        }

        val objects = layer.objects

        for (`object` in objects) {
            if (`object` is TextureMapObject) {
                continue
            }

            val properties = `object`.properties
            val type = properties.get("type") as String
            if (type == "zombieCuasy" || type == "zombieFrank" || type == "zombieMummy" || type == "zombieKid"
                || type == "zombiePan"
            ) {
                if (`object` is RectangleMapObject) {
                    createZombie(`object`, type)
                    worldGame!!.TOTAL_ZOMBIES_LEVEL++
                }
            } else {
                throw GdxRuntimeException("Error en layer " + "malos" + ", objeto:" + type)
            }
        }
    }

    private fun createPlatform(`object`: MapObject) {
        val rectangle = (`object` as RectangleMapObject).rectangle
        val polygon = PolygonShape()
        val size = Vector2((rectangle.x + rectangle.width * 0.5f) * m_units, (rectangle.y + rectangle.height * 0.5f) * m_units)
        polygon.setAsBox(rectangle.getWidth() * 0.5f * m_units, rectangle.height * 0.5f * m_units, size, 0.0f)
        defaultFixtureDefinition.shape = polygon

        val bodyDef = BodyDef()
        bodyDef.type = BodyType.StaticBody
        val body = world.createBody(bodyDef)
        body.createFixture(defaultFixtureDefinition)

        val x = (rectangle.x + rectangle.width * 0.5f) * m_units
        val y = (rectangle.y + rectangle.height * 0.5f) * m_units
        val height = (rectangle.height * m_units * 0.5f)
        val width = (rectangle.width * m_units * 0.5f)
        body.userData = Platform(x, y, width, height)
    }

    private fun createLadder(`object`: MapObject, tipo: String?) {
        val rectangle = (`object` as RectangleMapObject).rectangle
        val polygon = PolygonShape()
        val size = Vector2((rectangle.x + rectangle.width * 0.5f) * m_units, (rectangle.y + rectangle.height * 0.5f) * m_units)
        polygon.setAsBox(rectangle.getWidth() * 0.5f * m_units, rectangle.height * 0.5f * m_units, size, 0.0f)
        defaultFixtureDefinition.shape = polygon

        val bodyDef = BodyDef()
        bodyDef.type = BodyType.StaticBody
        val body = world.createBody(bodyDef)

        defaultFixtureDefinition.isSensor = true
        body.createFixture(defaultFixtureDefinition)
        body.userData = tipo

        defaultFixtureDefinition.isSensor = false
    }

    private fun createCrate(`object`: MapObject) {
        val rectangle = (`object` as RectangleMapObject).rectangle

        val polygon = PolygonShape()
        val width = (rectangle.width * m_units)
        val x = (rectangle.x + rectangle.width * 0.5f) * m_units
        val y = (rectangle.y + rectangle.height * 0.5f) * m_units

        polygon.setAsBox(width / 2f, width / 2f)

        val fixDef = FixtureDef()
        fixDef.density = 12f
        fixDef.friction = .7f
        fixDef.restitution = 0.0f
        fixDef.shape = polygon

        val obj = Crate(x, y, width)

        val bodyDefinition = BodyDef()
        bodyDefinition.position.x = obj.position.x
        bodyDefinition.position.y = obj.position.y
        bodyDefinition.type = BodyType.DynamicBody
        val body = world.createBody(bodyDefinition)

        body.createFixture(fixDef)

        worldGame!!.crates!!.add(obj)
        body.userData = obj
    }

    private fun createSaw(`object`: MapObject) {
        val rectangle = (`object` as RectangleMapObject).rectangle

        val width = (rectangle.width * m_units)
        val x = (rectangle.x + rectangle.width * 0.5f) * m_units
        val y = (rectangle.y + rectangle.height * 0.5f) * m_units

        val shape = CircleShape()
        shape.radius = width / 2f

        defaultFixtureDefinition.shape = shape

        val saw = Saw(x, y, width)

        val bodyDef = BodyDef()
        bodyDef.position.x = saw.position.x
        bodyDef.position.y = saw.position.y
        bodyDef.type = BodyType.KinematicBody
        val body = world.createBody(bodyDef)

        body.createFixture(defaultFixtureDefinition)

        worldGame!!.saws!!.add(saw)
        body.userData = saw
        body.angularVelocity = Math.toRadians(360.0).toFloat()

        shape.dispose()
    }

    private fun createItem(`object`: MapObject, type: String) {
        var obj: Items? = null

        val rectangle = (`object` as RectangleMapObject).rectangle
        val x = (rectangle.x + rectangle.width * 0.5f) * m_units
        val y = (rectangle.y + rectangle.height * 0.5f) * m_units

        when (type) {
            "gem" -> obj = ItemGem(x, y)
            "heart" -> {
                obj = ItemHeart(x, y)
                Gdx.app.log("Heart", "")
            }

            "skull" -> obj = ItemSkull(x, y)
            "meat" -> obj = ItemMeat(x, y)
            "shield" -> obj = ItemShield(x, y)
            "star" -> obj = ItemStar(x, y)
        }

        val bodyDefinition = BodyDef()
        bodyDefinition.position.y = obj!!.position.y
        bodyDefinition.position.x = obj.position.x
        bodyDefinition.type = BodyType.StaticBody

        val shape = CircleShape()
        shape.radius = .15f

        val fixtureDefinition = FixtureDef()
        fixtureDefinition.shape = shape
        fixtureDefinition.isSensor = true

        val body = world.createBody(bodyDefinition)
        body.createFixture(fixtureDefinition)

        body.userData = obj
        worldGame!!.items.add(obj)
        shape.dispose()
    }

    private fun createZombie(`object`: MapObject, type: String) {
        var obj: Zombie? = null

        val rectangle = (`object` as RectangleMapObject).rectangle
        val x = (rectangle.x + rectangle.width * 0.5f) * m_units
        val y = (rectangle.y + rectangle.height * 0.5f) * m_units

        when (type) {
            "zombieCuasy" -> obj = Zombie(x, y, Zombie.TYPE_CUASY)
            "zombieFrank" -> obj = Zombie(x, y, Zombie.TYPE_FRANK)
            "zombieKid" -> obj = Zombie(x, y, Zombie.TYPE_KID)
            "zombieMummy" -> obj = Zombie(x, y, Zombie.TYPE_MUMMY)
            "zombiePan" -> obj = Zombie(x, y, Zombie.TYPE_PAN)
        }

        val bodyDefinition = BodyDef()
        bodyDefinition.position.x = obj!!.position.x
        bodyDefinition.position.y = obj.position.y
        bodyDefinition.type = BodyType.DynamicBody

        val body = world.createBody(bodyDefinition)

        val shape = PolygonShape()
        shape.setAsBox(.17f, .32f)

        val fixtureDefinition = FixtureDef()
        fixtureDefinition.shape = shape
        fixtureDefinition.density = 8f
        fixtureDefinition.friction = 0f
        fixtureDefinition.filter.groupIndex = -1
        body.createFixture(fixtureDefinition)

        body.isFixedRotation = true
        body.userData = obj
        worldGame!!.zombies.add(obj)

        shape.dispose()
    }

    private fun getRectangle(rectangleObject: RectangleMapObject): Shape {
        val rectangle = rectangleObject.rectangle
        val polygon = PolygonShape()
        val size = Vector2((rectangle.x + rectangle.width * 0.5f) * m_units, (rectangle.y + rectangle.height * 0.5f) * m_units)
        polygon.setAsBox(rectangle.getWidth() * 0.5f * m_units, rectangle.height * 0.5f * m_units, size, 0.0f)
        return polygon
    }

    private fun getCircle(circleObject: CircleMapObject): Shape {
        val circle = circleObject.circle
        val circleShape = CircleShape()
        circleShape.radius = circle.radius * m_units
        circleShape.position = Vector2(circle.x * m_units, circle.y * m_units)
        return circleShape
    }

    private fun getPolygon(polygonObject: PolygonMapObject): Shape {
        val polygon = PolygonShape()
        val vertices = polygonObject.polygon.vertices
        val worldVertices = FloatArray(vertices.size)
        val yLost = polygonObject.polygon.y * m_units
        val xLost = polygonObject.polygon.x * m_units

        for (i in vertices.indices) {
            if (i % 2 == 0) worldVertices[i] = vertices[i] * m_units + xLost
            else worldVertices[i] = vertices[i] * m_units + yLost
        }
        polygon.set(worldVertices)

        return polygon
    }

    private fun getPolyline(polylineObject: PolylineMapObject): Shape {
        val vertices = polylineObject.polyline.vertices

        val worldVertices = arrayOfNulls<Vector2>(vertices.size / 2)
        val yLost = polylineObject.polyline.y * m_units
        val xLost = polylineObject.polyline.x * m_units
        for (i in 0..<vertices.size / 2) {
            worldVertices[i] = Vector2()
            worldVertices[i]!!.x = vertices[i * 2] * m_units + xLost
            worldVertices[i]!!.y = vertices[i * 2 + 1] * m_units + yLost
        }
        val chain = ChainShape()
        chain.createChain(worldVertices)
        return chain
    }
}
